docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+
+
+A setter which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden setter parameter type.
+
+
+A field which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden setter parameter type if this type is
+the same as the return type of the overridden getter (if the types are not the
+same then inference fails with an error).
+
+
+Note that overriding a field is addressed via the implicit induced getter/setter
+pair (or just getter in the case of a final field).
+
+
+A getter or setter with no annotated type that does not override anything is
+treated as if it were annotated with dynamic.
+
+
+A field with no annotated type that does not override anything has the type
+inferred from its initializer.
+
+
+### Initializer inference
+
+
+Initializer inference works by defining a subset of the expressions in the
+language for which we can predictably do inference in a way that is not
+sensitive to ordering outside of the static dependencies that are syntactically
+apparent in the expression.  In particular, field inference cannot change the
+outcome of initializer inference.
+
+
+A meta-goal is to keep this inference relatively lightweight and efficient to
+implement.
+
+
+In all cases, if initializer inference fails, it is an error and produces no
+result.  Implementations are free to treat this as a result of dynamic after
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Immediately evident expressions
+
+We define a set of expressions with immediately-evident types (called
+immediately-evident expressions, abbreviated IE) as follows.  Immediately
+evident expressions have an inferred type as specified below.  Any expression
+which is not an IE has no inferred type.  For every immediately evident
+expression, we also define the set of variable dependencies upon which its
+inference relies.
+
+
+* null, boolean, numeric, string, and symbol literals have their corresponding
+   type.
+   * No inference dependencies
+* A list literal or a map literal with explicit type arguments has the type
+   `List<T>` or `Map<K, V>` respectively, where `<T>` or `<K, V>` are the
+   provided type arguments.
+   * No inference dependencies
+* A non-empty list literal with no type arguments, and for which all of the
+   elements are IEs, has the type `List<T>` where `T` is the least upper bound
+   of the inferred types of the elements.
+   * If the elements do not all have an inferred type, it is an error
+   * The inference dependencies of the list literal are the collected
+      dependencies of the elements.
+* An non-const empty list literal with no type arguments has the type
+  `List<dynamic>`.
+   * No inference dependencies.
+* A const empty list literal with no type arguments has the type
+  `List<bottom>`.
```

Comment:
Unfortunately, we'll probably have to use the notation `List<Null>` for that, for some time.